### ISSUE TITLE
Help us prepare for the next major release of stylelint

### ISSUE THREAD

**Issue by jeddy3** [kristerkari/stylelint-scss#541]
We've [been working on](https://github.com/stylelint/stylelint/issues/5205) a new major release of stylelint. There are [quite a few changes](https://github.com/stylelint/stylelint/blob/v14/CHANGELOG.md#head) (and a [lot of refactoring](https://github.com/stylelint/stylelint/pulls?q=is%3Apr+refactor+)). One of the biggest changes we've made is [removing the `syntax` option](https://github.com/stylelint/stylelint/issues/5289#issuecomment-840772900) and no longer bundling syntaxes in stylelint itself. Users will need to extend stylelint (via shared-configs, plugins and custom syntaxes) to lint something other than vanilla CSS.

To make this easier for users to do, we've added two properties to the configuration object:
- [`customSyntax`](https://github.com/stylelint/stylelint/pull/5538) - which accepts an PostCSS syntax like [`postcss-scss`](https://github.com/postcss/postcss-scss)
- [`overrides`](https://github.com/stylelint/stylelint/pull/5521) - which allows users to configure stylelint to lint two or more languages, e.g. CSS and SCSS

We [intend to improve the migration experience](https://github.com/stylelint/stylelint/issues/5301) by updating our docs and by adding helpful error messages to stylelint when a user, for example, tries to lint an `*.scss` file but hasn't specified a custom syntax in their configuration object.

This last part is [why we're reaching out to you](https://github.com/stylelint/stylelint/issues/5205#issuecomment-856669471). We want to encourage users to extend a shared-config for the language they are linting, rather than add custom syntaxes and plugins themselves.

We also think emphasising the [`standard`](https://github.com/stylelint/stylelint-config-standard) configs over the [`recommended`](https://github.com/stylelint/stylelint-config-recommended) ones will help users, especially new ones. Including those who use stylelint and a pretty-printer like Prettier together, as there are a number of stylistic rules in stylelint and this plugin that are complementary to Prettier, e.g. the `*-empty-line-before` ones.

To help us do this, will you (or any of the other contributors to this package) be willing to?:

1. Update [`stylelint-config-recommended-scss`](https://github.com/kristerkari/stylelint-config-recommended-scss) to bundle [`postcss-scss`](https://github.com/postcss/postcss-scss)
2. Create and publish `stylelint-config-standard-scss` that would:
   - extend both [`stylelint-config-standard`](https://github.com/stylelint/stylelint-config-standard) and [`stylelint-config-recommended-scss`](https://github.com/kristerkari/stylelint-config-recommended-scss)
   - configure stylelint's stylistic rules for SCSS
   - turn on and configure all the stylistic rules in this plugin

```js
// stylelint-config-recommended-scss
module.exports = {
  extends: ["stylelint-config-recommended"],
  plugins: ["stylelint-scss"],
  customSyntax: "postcss-scss",
  rules: {
    "at-rule-no-unknown": null,
    "scss/at-rule-no-unknown": true,
    "no-invalid-position-at-import-rule": null
  }
};
```

```js
// stylelint-config-standard-scss
module.exports = {
  extends: [
    "stylelint-config-standard",
    "stylelint-config-recommended-scss"
  ],
  rules: {
    "block-closing-brace-newline-after": ["always", { "ignoreAtRules": ["else"] } ],
    "scss/at-else-closing-brace-newline-after": "always-last-in-chain"
    ..
  }
};
```

We'll then update our [Getting started guide](https://stylelint.io/user-guide/get-started) and add errors messages to stylelint to encourage users to install `stylelint-config-standard-scss` when linting SCSS.

We've seen users linting SCSS with stylelint who are unaware of this plugin and all the amazing rules within it. We hope these changes will allow more users to benefit from this plugin. 

Lastly, many thanks to you (and the other contributors) for maintaining this package! It's been fantastic to see the number of rules grow over the years.

**Comment by kristerkari** [comment]
Thanks @jeddy3! Looks like there are quite a few things that need to get done. 

The syntax option change sounds like a big breaking change, so probably the SCSS related packages need a major version bump.

**Comment by niksy** [comment]
> The syntax option change sounds like a big breaking change, so probably the SCSS related packages need a major version bump.

I think it definitely should be major version bump since PostCSS 8 is used internally in new Stylelint version.

@kristerkari what can we do to help with this?

**Comment by jeddy3** [comment]
> The syntax option change sounds like a big breaking change

Yes, it's not a decision we took lightly. It mainly affects users, not plugin authors. We hope to make the migration for users as smooth as possible by suggesting they extend shared configs that bundle the appropriate syntax for them.

> Looks like there are quite a few things that need to get done.

I believe there are two user-facing things the SCSS community need to do:
- add the `customSyntax` property to `stylelint-config-recommended-scss`
- create `stylelint-config-standard-scss`

> so probably the SCSS related packages need a major version bump.

This plugin shouldn't need any user-facing changes and should already be compatible with the upcoming release of stylelint. We migrated to PostCSS 8 internally, but plugins are [made compatible by stylelint itself](https://github.com/stylelint/stylelint/pull/5353).

Adding the `customSyntax` to `stylelint-config-recommended-scss` is also non-breaking as the current version of stylelint will silently ignore the property.

> @kristerkari what can we do to help with this?

There are quite a few moving parts. It would be good to get some pull requests open so that we can see the parts working together. I suggest:
- create pull request to add `customSyntax` to `stylelint-config-recommended-scss`
- create a `stylelint-config-standard-scss` repo and add a pull request to:
  - extend [this branch](https://github.com/stylelint/stylelint-config-standard/pull/198) of [`stylelint-config-standard`](https://github.com/stylelint/stylelint-config-standard) and the above branch of [`stylelint-config-recommended-scss`](https://github.com/kristerkari/stylelint-config-recommended-scss)
   - configure stylelint's stylistic rules for SCSS
   - turn on and configure all the stylistic rules in this plugin
- create a pull request in this repo to update the stylelint dev dependency

All pull requests can use the [`v14` branch of stylelint](https://github.com/stylelint/stylelint/tree/v14) for now.

We'd then be able to test the upcoming `stylelint-config-recommended-scss` package, which under the hood will test:
- the upcoming release of stylelint (with its `customSyntax` option) 
- the upcoming release of `stylelint-config-standard`

Updating the stylelint dev dependency in this package will require tweaking the tests. You can use the stylelint jest preset (see https://github.com/kristerkari/stylelint-scss/pull/470) and then replace `syntax: "scss"` with `customSyntax: "postcss-scss"` in your tests files. However, you work around the issue more easily by requiring `postcss-scss` in your jest setup script by updating [this line](https://github.com/kristerkari/stylelint-scss/blob/master/jest-setup.js#L41]) to `customSyntax: require("postcss-scss")`.

Feel free to at-mention me in these pull requests, and I'll be happy to take a look.

@kristerkari Do those suggestions sound good to you? It looks like @niksy would be happy to get the ball rolling.

**Comment by jeddy3** [comment]
@niksy Fantastic work on https://github.com/kristerkari/stylelint-scss/pull/543 & https://github.com/kristerkari/stylelint-config-recommended-scss/pull/83. I'm going offline for a week now, but I'm happy to continue helping out when I'm back. In the meantime, if you like you can get the ball rolling on creating a `stylelint-config-standard-scss` repo as I'm sure there'll be some discussion around:
- how the [`stylelint-config-standard`](https://github.com/stylelint/stylelint-config-standard) rules should be configured for SCSS
- which rules in this plugin should be turned on and how they should be configured

For context, [`stylelint-config-standard`](https://github.com/stylelint/stylelint-config-standard) enforces:
- the most common conventions found in the examples of the specifications
- common patterns found in a handful of popular CSS styleguides

Perhaps, something similar exists for SCSS, e.g. examples in the SCSS Handbook and popular SCSS styleguides?

In your pull request, I suggest using the branches of these pull requests as your dependencies:
- https://github.com/kristerkari/stylelint-config-recommended-scss/pull/83
- https://github.com/stylelint/stylelint-config-standard/pull/198

i.e.:

- `github:niksy/stylelint-config-recommended-scss/#stylelint-v14`
- `github:stylelint/stylelint-config-recommended#add-rules`

And the `v14` branch of stylelint. You'll then be working with the latest of everything.

**Comment by niksy** [comment]
@jeddy3 

> I'm going offline for a week now, but I'm happy to continue helping out when I'm back.

Happy vacationing! :)

> In the meantime, if you like you can get the ball rolling on creating a `stylelint-config-standard-scss` repo as I'm sure there'll be some discussion around

@kristerkari can you create that repo so we can get started? Sass related stuff is under your username so it’s best to keep it there. Or maybe we can create `stylelint-scss` GitHub organization?

**Comment by kristerkari** [comment]
> @kristerkari can you create that repo so we can get started? Sass related stuff is under your username so it’s best to keep it there. Or maybe we can create `stylelint-scss` GitHub organization?

I'm not sure if it makes sense to add more stuff under my name, it would prefer to start adding stuff under `stylelint` org or then create a new `stylelint-scss` org.

**Comment by jeddy3** [comment]
A `stylelint-scss` org sounds good to me as the `stylelint` org is limited to standard CSS syntax.

@niksy I say go ahead, create it and start working on the `standard` config.

**Comment by jeddy3** [comment]
I'm around this week and can help out. Just at-mention me in any repos.
